File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        int n = rains.size();
        vector<int> result(n, -1);
        unordered_map<int, int> lakeFullDay;  // lake -> day it was filled
        set<int> dryDays;  // sorted set of available dry days
        
        for (int i = 0; i < n; i++) {
            if (rains[i] == 0) {
                // Dry day - we'll decide which lake to dry later
                dryDays.insert(i);
                result[i] = 1;  // Default: dry lake 1 (will be overwritten if needed)
            } else {
                int lake = rains[i];
                if (lakeFullDay.count(lake)) {
                    // Lake already full - need to find a dry day after it was filled
                    int lastFullDay = lakeFullDay[lake];
                    auto it = dryDays.upper_bound(lastFullDay);
                    
                    if (it == dryDays.end()) {
                        // No available dry day - flood is inevitable
                        return {};
                    }
                    
                    // Use this dry day to dry the lake
                    result[*it] = lake;
                    dryDays.erase(it);
                }
                // Mark lake as full on this day
                lakeFullDay[lake] = i;
            }
        }
        return result;
    }
};

File: 1612-avoid-flood-in-the-city/avoid-flood-in-the-city.py
from Queue import PriorityQueue

class Solution(object):
    def binarySearch(self, val, array):
        i = 0
        j = len(array) - 1
        
        while ( i <= j):
            mid = i + (j - i) /2
            
            if array[mid] > val and (mid == 0 or array[mid - 1] <= val):
                return mid
            
            if array[mid] > val:
                j = mid - 1
            else:
                i = mid + 1
        return None
            
    def avoidFlood(self, rains):
        
        lakeRain = dict()
        dryDays = []
        result = []

        for day, lake in enumerate(rains):
            if lake == 0:
                dryDays.append(day)
                result.append(1)
            else:
                if lake not in lakeRain:
                    lakeRain[lake] = day
                    result.append(-1)
                else:
                    result.append(-1)
                    chosenDay = None
                    previousRainDay = lakeRain[lake]
                    
                    chosenDayIndex = self.binarySearch(previousRainDay, dryDays)
                    
                    if chosenDayIndex is None:
                        return []
                    else:
                        dryDay = dryDays[chosenDayIndex]
                        dryDays.pop(chosenDayIndex)
                        lakeRain[lake] = day
                        result[dryDay] = lake
        return result